cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java

```diff
@@ -0,0 +1,179 @@
+ * the License.
+ */
+
+package io.cdap.cdap.internal.credential.store;
+
+import com.google.gson.Gson;
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.common.AlreadyExistsException;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.proto.credential.CredentialIdentity;
+import io.cdap.cdap.proto.id.CredentialIdentityId;
+import io.cdap.cdap.proto.id.CredentialProfileId;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.StoreDefinition.CredentialProvisionerStore;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Storage for credential identities.
+ */
+public class CredentialIdentityStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CredentialIdentityStore.class);
+  private static final Gson GSON = new Gson();
+
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  CredentialIdentityStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  public Collection<CredentialIdentityId> list(String namespace) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_IDENTITIES);
+      Field<?> namespaceIndex = Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+          namespace);
+      try (CloseableIterator<StructuredRow> iterator = table.scan(Range.singleton(
+          Collections.singleton(namespaceIndex)), Integer.MAX_VALUE)) {
+        return identitiesFromRowIterator(iterator);
+      }
+    }, IOException.class);
+  }
+
+  public Optional<CredentialIdentity> get(CredentialIdentityId id) throws IOException {
+    Utils.validateResourceName(id.getName());
+    return TransactionRunners.run(transactionRunner, context -> {
+      Optional<StructuredRow> row = fetchRow(context, id);
+      if (!row.isPresent()) {
+        return Optional.empty();
+      }
+      return Optional.of(GSON.fromJson(row.get()
+          .getString(CredentialProvisionerStore.IDENTITY_DATA_FIELD), CredentialIdentity.class));
+    }, IOException.class);
+  }
+
+  public void create(CredentialIdentityId id, CredentialIdentity identity)
```

Comment:
will we have separate create and update APIs? most of our REST APIs just have a PUT, which ends up being create or update in the same call.

Reply:
I'm keeping the APIs separate because it adds simplicity for the access control logic, otherwise we'll need to check different permissions depending on if the resource exists or not. It's possible to write the logic that way, but seems unnecessarily complex IMO. WDYT?